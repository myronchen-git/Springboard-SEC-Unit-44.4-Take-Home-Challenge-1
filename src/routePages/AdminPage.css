.AdminPage {
  width: 90%;

  text-align: center;
}

.AdminPage table {
  box-shadow: 0 0 0.75em grey;
}
