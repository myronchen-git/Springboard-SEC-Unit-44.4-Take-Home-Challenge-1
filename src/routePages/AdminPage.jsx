import { useContext, useEffect, useState } from 'react';
import { Table } from 'reactstrap';

import { UserContext } from '../contexts';

import './AdminPage.css';

// ==================================================

/**
 * Displays the administration page, which just shows the list of users.
 */
function AdminPage() {
  const [users, setUsers] = useState([]);
  const [errorMessage, setErrorMessage] = useState(null);
  const { getUsers } = useContext(UserContext);

  useEffect(() => {
    getUsers()
      .then((data) => setUsers(data))
      .catch((err) => setErrorMessage(err.message));
  }, [getUsers]);

  if (errorMessage) {
    return (
      <main className="AdminPage">
        <title>Yodlr Admin Page</title>
        <h1>Users</h1>
        <div>
          <h1>Error</h1>
          <p>{errorMessage}</p>
        </div>
      </main>
    );
  } else {
    return (
      <main className="AdminPage">
        <title>Yodlr Admin Page</title>
        <h1>Users</h1>
        <Table striped>
          <thead>
            <tr>
              <th>ID</th>
              <th>First Name</th>
              <th>Last Name</th>
              <th>Email</th>
              <th>State</th>
            </tr>
          </thead>
          <tbody>
            {users.map((user) => (
              <tr key={user.id}>
                <td>{user.id}</td>
                <td>{user.firstName}</td>
                <td>{user.lastName}</td>
                <td>{user.email}</td>
                <td>{user.state}</td>
              </tr>
            ))}
          </tbody>
        </Table>
      </main>
    );
  }
}

// ==================================================

export default AdminPage;
