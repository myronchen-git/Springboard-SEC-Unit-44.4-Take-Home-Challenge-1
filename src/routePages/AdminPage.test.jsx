import { act, render } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import { UserContext } from '../contexts.jsx';
import AdminPage from './AdminPage';

// ==================================================

describe('AdminPage', () => {
  const users = Object.freeze([
    Object.freeze({
      id: 1,
      firstName: 'First Name 1',
      lastName: 'Last Name 1',
      email: 'email1@email.com',
    }),
    Object.freeze({
      id: 2,
      firstName: 'First Name 2',
      lastName: 'Last Name 2',
      email: 'email2@email.com',
    }),
  ]);

  const mockGetUsers = vi.fn();

  beforeEach(() => {
    mockGetUsers.mockReset();
    mockGetUsers.mockResolvedValue(users);
  });

  it('renders.', async () => {
    await act(async () =>
      render(
        <UserContext.Provider value={{ getUsers: mockGetUsers }}>
          <AdminPage />
        </UserContext.Provider>
      )
    );
  });

  it('matches snapshot.', async () => {
    const { asFragment } = await act(async () =>
      render(
        <UserContext.Provider value={{ getUsers: mockGetUsers }}>
          <AdminPage />
        </UserContext.Provider>
      )
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('lists all users.', async () => {
    // Act
    const { getByText } = await act(async () =>
      render(
        <UserContext.Provider value={{ getUsers: mockGetUsers }}>
          <AdminPage />
        </UserContext.Provider>
      )
    );

    // Assert
    for (const user of users) {
      for (const userInfo of Object.values(user)) {
        expect(getByText(userInfo)).toBeVisible();
      }
    }
  });
});
