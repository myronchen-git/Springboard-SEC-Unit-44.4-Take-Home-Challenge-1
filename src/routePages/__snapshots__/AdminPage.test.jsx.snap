// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AdminPage > matches snapshot. 1`] = `
<DocumentFragment>
  <main
    class="AdminPage"
  >
    <title>
      Yodlr Admin Page
    </title>
    <h1>
      Users
    </h1>
    <table
      class="table table-striped"
    >
      <thead>
        <tr>
          <th>
            ID
          </th>
          <th>
            First Name
          </th>
          <th>
            Last Name
          </th>
          <th>
            Email
          </th>
          <th>
            State
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            1
          </td>
          <td>
            First Name 1
          </td>
          <td>
            Last Name 1
          </td>
          <td>
            email1@email.com
          </td>
          <td />
        </tr>
        <tr>
          <td>
            2
          </td>
          <td>
            First Name 2
          </td>
          <td>
            Last Name 2
          </td>
          <td>
            email2@email.com
          </td>
          <td />
        </tr>
      </tbody>
    </table>
  </main>
</DocumentFragment>
`;
